### DISCUSSION TITLE
Kubernetes Production / Staging Deployment and When to actually Build Next?

### DISCUSSION THREAD

**Discussion by sherifabdlnaby** [vercel/next.js#25176] [Help]
Hello, I come from the DevOps/SRE world, meaning I don't write NextJS or JS in my day-to-day job. I've been studying NextJS to make our production deployment as optimized and performant as possible.

I want to open a discussion about how do you deploy NextJS in production using Kubernetes.

# Where do you Build your Next App?

When building the container image? Or Upon starting the application container?

#### Pros
1. Container startup time is very fast.
2. Build Once, Run Many containers replicas.
3. Containers' resources/limits are given based on the application runtime needs, as typically build requires much more resources than serving actual content.

#### Cons
1. Build can't be parameterized; We have multiple QA and Staging Enviroments, this means URLs like `NEXT_PUBLIC_BACKEND_API` need to change "dynamically". If we run `next build` at image build-time, we won't be able to parameterize these configuration based on where it is deployed.
2. Image building is done on an isolated cluster that may not be able to reach actual backend servers, as `getStaticPropes` will run at build time and may need to access a server that is not reachable at the image building time. Alternatively, I'll need to build an Image for each environment which is infeasible since we have dynamic custom environments.

#### Pros
1. Build run on the same environment the NextJS server will run on, meaning it will be able to reach all its backend APIs.
2. Build can be parameterized per-environment configuration.

#### Cons
1. Huge startup time for the container.
2. Container has to be scheduled with enough resources to run the build, even if they'll use only a fraction of these resources to serve the content. 

I may not be aware of all of NextJS deployment methods and optimization, I want to know more about people's experiences.

**Comment by realmunk** [comment]
Did you ever manage to get some help with this, @sherifabdlnaby? We're seeing that multiple of the questions regarding self-hosting have been ignored so far.Would love some clarity into what the recommended process is.
